### ISSUE TITLE
segment crash when deleting large number of duplicate rows from table using ctid, gp_segment_id 

### ISSUE THREAD

**Issue by pf-qiu** [greenplum-db/gpdb#9941]
For a table with a logical primary key which is not enforced by gpdb, user wants to remove duplicate rows from the table.
```
DELETE FROM t 
USING (
    SELECT max(ctid) AS ctid, gp_segment_id, pk
    FROM t GROUP BY (gp_segment_id, pk)
    HAVING count(*) > 1
) tt
WHERE t.pk = tt.pk
AND t.gp_segment_id = tt.gp_segment_id
AND t.ctid <> tt.ctid;
```
On master this will crash if table has more than 1 million duplicate rows per segment. 5X and 6X don't have the same symptom.

### Greenplum version or build
PostgreSQL 9.6beta4 (Greenplum Database 7.0.0-alpha.0+dev.6742.gb5b28dec73 build dev) on x86_64-unknown-linux-gnu, compiled by gcc (GCC) 6.2.0, 64-bit compiled on Apr 16 2020 09:02:0
9
### OS version and uname -a
Linux 9e2ecf7ddcd8 5.3.0-46-generic #38~18.04.1-Ubuntu SMP Tue Mar 31 04:17:56 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
### autoconf options used ( config.status --config )
'--prefix=/home/gpadmin/gp' '--disable-orca' '--disable-gpcloud' '--disable-pxf' '--enable-debug' 'CFLAGS=-O0'
### Expected behavior
DELETE 3000000
### Actual behavior
ERROR:  Error on receive from seg1 172.17.0.2:6001 pid=10032: server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
### Step to reproduce the behavior
For an N segments cluster, generate_series(1, N*1000000) will do.
```
CREATE TABLE t(pk int, c int) DISTRIBUTED BY(pk);
INSERT INTO t SELECT x,x FROM generate_series(1, 3000000) x;
INSERT INTO t SELECT x,x FROM generate_series(1, 3000000) x;

DELETE FROM t 
USING (
    SELECT max(ctid) AS ctid, gp_segment_id, pk
    FROM T GROUP BY (gp_segment_id, pk)
    HAVING count(*) > 1
) tt
WHERE t.pk = tt.pk
AND t.gp_segment_id = tt.gp_segment_id
AND t.ctid <> tt.ctid;
```
Part of stack trace:
```
#0  0x00000000037b01e0 in ?? ()
#1  0x0000000000af0f06 in pfree (pointer=0x37b0260) at mcxt.c:1197
#2  0x0000000000739d7f in advance_combine_function (aggstate=0x2c0d498, pertrans=0x2c17288, pergroupstate=0x37b0250, fcinfo=0x2c17ef0) at nodeAgg.c:1038
#3  0x000000000077a3bc in agg_hash_reload (aggstate=0x2c0d498) at execHHashagg.c:1957
#4  0x000000000077ab98 in agg_hash_next_pass (aggstate=0x2c0d498) at execHHashagg.c:2189
#5  0x000000000073b643 in agg_retrieve_hash_table (aggstate=0x2c0d498) at nodeAgg.c:2161
#6  0x000000000073af0c in ExecAgg (node=0x2c0d498) at nodeAgg.c:1769
#7  0x00000000007218b5 in ExecProcNode (node=0x2c0d498) at execProcnode.c:729
#8  0x000000000075f012 in SubqueryNext (node=0x2c08a88) at nodeSubqueryscan.c:55
#9  0x000000000072ee08 in ExecScanFetch (node=0x2c08a88, accessMtd=0x75eff3 <SubqueryNext>, recheckMtd=0x75f01c <SubqueryRecheck>) at execScan.c:97
#10 0x000000000072ef3b in ExecScan (node=0x2c08a88, accessMtd=0x75eff3 <SubqueryNext>, recheckMtd=0x75f01c <SubqueryRecheck>) at execScan.c:167
#11 0x000000000075f051 in ExecSubqueryScan (node=0x2c08a88) at nodeSubqueryscan.c:87
#12 0x00000000007217a4 in ExecProcNode (node=0x2c08a88) at execProcnode.c:671
#13 0x00000000007471c2 in ExecHashJoinOuterGetTuple (outerNode=0x2c08a88, hjstate=0x2c09768, hashvalue=0x7ffcf529bea4) at nodeHashjoin.c:820
#14 0x00000000007463ca in ExecHashJoin_guts (node=0x2c09768) at nodeHashjoin.c:287
#15 0x0000000000746a35 in ExecHashJoin (node=0x2c09768) at nodeHashjoin.c:514
#16 0x0000000000721876 in ExecProcNode (node=0x2c09768) at execProcnode.c:714
#17 0x0000000000752b76 in ExecModifyTable (node=0x2c09528) at nodeModifyTable.c:1894
#18 0x000000000072167e in ExecProcNode (node=0x2c09528) at execProcnode.c:605
#19 0x000000000071c8db in ExecutePlan (estate=0x2c08240, planstate=0x2c09528, use_parallel_mode=0 '\000', operation=CMD_DELETE, sendTuples=0 '\000', numberTuples=0, 
    direction=ForwardScanDirection, dest=0x2c062f8) at execMain.c:2826
#20 0x0000000000719a72 in standard_ExecutorRun (queryDesc=0x2b529e0, direction=ForwardScanDirection, count=0) at execMain.c:852
```

**Comment by hlinnaka** [comment]
This looks like the same issue as https://github.com/greenplum-db/gpdb/issues/9916

**Comment by pf-qiu** [comment]
Yeah. I will test this again if #9916 is resolved.

**Comment by hlinnaka** [comment]
#9916 has been fixed on master. However, I could not reproduce this even when I reverted commit 736898adbb90f8fc9691d639ba3e030a21e4c1b8, so I'm not sure I can test this properly. Can you confirm that this has been fixed?

**Comment by pf-qiu** [comment]
Hi Heikki, indeed it's already fixed before 736898a.
Using bisect I figured out that this issue is fixed by d5be50d in https://github.com/greenplum-db/gpdb/pull/9929.
I will close this one.
